analysis/cpp/fast_em.cc

```diff
@@ -0,0 +1,294 @@
+// Copyright 2015 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <assert.h>
+#include <stdarg.h>  // va_list, etc.
+#include <stdio.h>  // fread()
+#include <stdlib.h>  // exit()
+#include <stdint.h>  // uint16_t
+#include <string.h>  // strcmp()
+#include <cmath>  // std::abs operates on doubles
+#include <cstdlib>  // strtol
+#include <string>
+#include <vector>
+
+using std::string;
+using std::vector;
+
+// Log messages to stdout.
+void log(const char* fmt, ...) {
+  va_list args;
+  va_start(args, fmt);
+  vprintf(fmt, args);
+  va_end(args);
+  printf("\n");
+}
+
+const int kTagLen = 4;  // 4 byte tags in the file format
+
+bool ExpectTag(FILE* f, const char* tag) {
+  char buf[kTagLen];
+
+  if (fread(buf, sizeof buf[0], kTagLen, f) != kTagLen) {
+    return false;
+  }
+  if (strcmp(buf, tag) != 0) {
+    log("Error: expected '%s'", tag);
+    return false;
+  }
+  return true;
+}
+
+static bool ReadListOfMatrices(
+    FILE* f, uint32_t* num_entries_out, uint32_t* entry_size_out,
+    vector<double>* v_out) {
+  if (!ExpectTag(f, "ne ")) {
+    return false;
+  }
+
+  // R integers are serialized as uint32_t
+  uint32_t num_entries;
+  if (fread(&num_entries, sizeof num_entries, 1, f) != 1) {
+    return false;
+  }
+
+  log("num entries: %d", num_entries);
+
+  if (!ExpectTag(f, "es ")) {
+    return false;
+  }
+
+  uint32_t entry_size;
+  if (fread(&entry_size, sizeof entry_size, 1, f) != 1) {
+    return false;
+  }
+  log("entry_size: %d", entry_size);
+
+  if (!ExpectTag(f, "dat")) {
+    return false;
+  }
+
+  // Now read dynamic data
+  uint32_t vec_length = num_entries * entry_size;
```

Comment:
You should really put bounds on these things so there is no possibility of wrapping here.